rcl/include/rcl/context.h

```diff
@@ -0,0 +1,251 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__CONTEXT_H_
+#define RCL__CONTEXT_H_
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include "rcutils/stdatomic_helper.h"
+#include "rmw/init.h"
+
+#include "rcl/allocator.h"
+#include "rcl/arguments.h"
+#include "rcl/init_options.h"
+#include "rcl/macros.h"
+#include "rcl/types.h"
+#include "rcl/visibility_control.h"
+
+typedef uint64_t rcl_context_instance_id_t;
+typedef atomic_uint_least64_t rcl_atomic_context_instance_id_t;
+
+struct rcl_context_impl_t;
+
+/// Encapsulates the non-global state of an init/shutdown cycle.
+/**
+ * The context is used in the creation of top level entities like nodes and
+ * guard conditions, as well as to shutdown a specific instance of init.
+ *
+ * Here is a diagram of a typical context's lifecycle:
+ *
+ * ```
+ *    +---------------+
+ *    |               |
+ * +--> uninitialized +---> rcl_get_zero_initialized_context() +
+ * |  |               |                                        |
+ * |  +---------------+                                        |
+ * |                                                           |
+ * |           +-----------------------------------------------+
+ * |           |
+ * |  +--------v---------+                +-----------------------+
+ * |  |                  |                |                       |
+ * |  | zero-initialized +-> rcl_init() +-> initialized and valid +-> rcl_shutdown() +
+ * |  |                  |                |                       |                  |
+ * |  +------------------+                +-----------------------+                  |
+ * |                                                                                 |
+ * |               +-----------------------------------------------------------------+
+ * |               |
+ * |  +------------v------------+
+ * |  |                         |
+ * |  | initialized but invalid +---> finalize all entities, then rcl_context_fini() +
+ * |  |                         |                                                    |
+ * |  +-------------------------+                                                    |
+ * |                                                                                 |
+ * +---------------------------------------------------------------------------------+
```

Comment:
Maybe using `\dot` or `\msc` would look nicer than ASCII art.

Reply:
dot is not very easy to understand as text, and I don't know what msc is. I won't spend more time on this diagram unless there's some other reason.

Reply:
I just looked up `msc`, which could be a useful way to display it in the rendered view, but as I said, I would prefer something that can be understood easily in plain text. If the diagram were more complex I'd give up on that and just use `dot` or something.